research/delf/delf/python/training/global_features_utils.py

```diff
@@ -0,0 +1,212 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for the global model training."""
+
+import os
+
+from absl import flags
+from absl import logging
+
+import tensorflow as tf
+import numpy as np
+
+from tensorboard import program
+
+from delf.python.datasets.revisited_op import dataset
+
+FLAGS = flags.FLAGS
+
+
+class AverageMeter():
+  """Computes and stores the average and current value of loss."""
+
+  def __init__(self):
+    """Initialization of the AverageMeter."""
+    self.reset()
+
+  def reset(self):
+    """Resets all the values."""
+    self.val = 0
+    self.avg = 0
+    self.sum = 0
+    self.count = 0
+
+  def update(self, val, n=1):
+    """Updates values in the AverageMeter.
+
+    Args:
+      val: Float, loss value.
+      n: Integer, number of instances.
+    """
+    self.val = val
+    self.sum += val * n
+    self.count += n
+    self.avg = self.sum / self.count
+
+
+def compute_metrics_and_print(dataset_name, sorted_index_ids, ground_truth,
+                              desired_pr_ranks=[1, 5, 10], log=True):
+  """Computes and loggs easy/medium/hard ground-truth metrics for Revisited
+  datasets.
+
+  Args:
+    dataset_name: String, name of the dataset.
+    sorted_index_ids: Integer NumPy array of shape [#queries, #index_images].
+      For each query, contains an array denoting the most relevant index images,
+      sorted from most to least relevant.
+    ground_truth: List containing ground-truth information for dataset. Each
+      entry is a dict corresponding to the ground-truth information for a query.
+      The dict has keys 'ok' and 'junk', mapping to a NumPy array of integers.
+    desired_pr_ranks: List of integers containing the desired precision/recall
+      ranks to be reported. E.g., if precision@1/recall@1 and
+      precision@10/recall@10 are desired, this should be set to [1, 10]. The
+      largest item should be <= #sorted_index_ids.
+
+  Returns:
+    mAP: [metricsE, metricsM, metricsH] List of the metrics for different
+      levels of complexity. Each metrics is a list containing:
+      mean_average_precision (float), mean_precisions (NumPy array of
+      floats, with shape [len(desired_pr_ranks)]), mean_recalls (NumPy array
+      of floats, with shape [len(desired_pr_ranks)]), average_precisions
+      (NumPy array of floats, with shape [#queries]), precisions (NumPy array of
+      floats, with shape [#queries, len(desired_pr_ranks)]), recalls (NumPy
+      array of floats, with shape [#queries, len(desired_pr_ranks)]).
+  """
+
+  if dataset_name.startswith('roxford5k') or dataset_name.startswith(
```

Comment:
seems like only Oxf/Par datasets are allowed? Maybe mention this in the docstring and raise a ValueError if it's not the case.